ftl/python/layer_builder.py

```diff
@@ -14,54 +14,227 @@
+            for f in os.listdir(self._wheel_dir)
+        ]
+
+    def _whl_to_fslayer(self, whl):
+        tmp_dir = tempfile.mkdtemp()
+        pkg_dir = os.path.join(tmp_dir, 'env')
+        os.makedirs(pkg_dir)
+
+        pip_cmd_args = list(self._pip_cmd)
+        pip_cmd_args.extend(['install', '--no-deps', '--prefix', pkg_dir, whl])
+
+        proc_pipe = subprocess.Popen(
+            pip_cmd_args,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            env=self._gen_pip_env(),
+        )
+        stdout, stderr = proc_pipe.communicate()
+        logging.info("`pip install` stdout:\n%s" % stdout)
+        if stderr:
+            logging.error("`pip install` had error output:\n%s" % stderr)
+        if proc_pipe.returncode:
+            raise Exception("error: `pip install` returned code: %d" %
+                            proc_pipe.returncode)
+        return tmp_dir
+
+    def _pip_install(self, pkg_txt):
+        with ftl_util.Timing('pip_download_wheels'):
+            pip_cmd_args = list(self._pip_cmd)
+            pip_cmd_args.extend(
+                ['wheel', '-w', self._wheel_dir, '-r', '/dev/stdin'])
+
+            proc_pipe = subprocess.Popen(
+                pip_cmd_args,
+                stdin=subprocess.PIPE,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.PIPE,
+                env=self._gen_pip_env(),
+            )
+            stdout, stderr = proc_pipe.communicate(input=pkg_txt)
+            logging.info("`pip wheel` stdout:\n%s" % stdout)
+            if stderr:
+                logging.error("`pip wheel` had error output:\n%s" % stderr)
+            if proc_pipe.returncode:
+                raise Exception("error: `pip wheel` returned code: %d" %
+                                proc_pipe.returncode)
+
+    def _gen_pip_env(self):
+        pip_env = os.environ.copy()
+        # bazel adds its own PYTHONPATH to the env
+        # which must be removed for the pip calls to work properly
+        pip_env.pop('PYTHONPATH', None)
+        pip_env['VIRTUAL_ENV'] = self._venv_dir
+        pip_env['PATH'] = self._venv_dir + '/bin' + ':' + os.environ['PATH']
+        return pip_env
+
+    def _log_cache_result(self, hit):
+        cache_str = constants.P1_CACHE_HIT_LOGSTR if hit else constants.P1_CACHE_MISS_LOGSTR
+        logging.info(cache_str.format(
+            key_version=constants.CACHE_KEY_VERSION,
+            language='PYTHON (requirements)',
+            key=self.GetCacheKey()
+        ))
 
 class InterpreterLayerBuilder(single_layer_image.CacheableLayerBuilder):
-    def __init__(self, venv_dir, python_cmd, venv_cmd):
+    def __init__(self, venv_dir=None, python_cmd=None,
+                 venv_cmd=None, cache=None):
         super(InterpreterLayerBuilder, self).__init__()
         self._venv_dir = venv_dir
         self._python_cmd = python_cmd
         self._venv_cmd = venv_cmd
+        self._cache = cache
 
     def GetCacheKeyRaw(self):
```

Comment:
might make sense to have the cache key value use the logging format or use the GetCacheKeyRaw values in the logging.  Currently something like CACHE_KEY_VERSION can be updated to help find the logs for updated cache keys but it might make more sense to have the version also embedded in the cache keys themselves.

Reply:
it isn't clear to me when to bump CACHE_KEY_VERSION, on any change made to a cache key?  it might be useful to couple them to not have to remember this

Reply:
yeah the main reason we need it is so if we ever change the key format, we know which log entries to ignore since the old ones will mess up our processing to construct the key mapping. I'll combine this with the other cache version variable we have for each language

Reply:
consolidated the keys and embedded the version in them.